File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int dfs(vector<int>& numbers, vector<vector<bool>>& visited, int target, int cur, int curNum) {
    // Base case
    if(cur == numbers.size()) {
        return curNum == target ? 1 : 0;
    }
    
    int ret = 0;
    // 음수일 경우
    if(!visited[cur][0]) {
        visited[cur][0] = true;
        ret += dfs(numbers, visited, target, cur + 1, curNum - numbers[cur]);
        visited[cur][0] = false;
    }
    
    // 양수일 경우
    if(!visited[cur][1]) {
        visited[cur][1] = true;
        ret += dfs(numbers, visited, target, cur + 1, curNum + numbers[cur]);
        visited[cur][1] = false;
    }
    
    return ret;
    
}

int solution(vector<int> numbers, int target) {
    // [원소 idx][0] -> 음수, [원소 idx][1] -> 양수
    vector<vector<bool>> visited(numbers.size(), vector<bool>(2, false));
    int result = 0;
    result = dfs(numbers, visited, target, 0, 0);
    return result;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    
    int[] numbers;
    int target;
    int answer = 0;
    int len;
    
    public int solution(int[] numbers, int target) {
        this.numbers = numbers;
        this.target = target;
        len = numbers.length;
        
        dfs(0, 0);
        return answer;
    }
    
    void dfs(int idx, int sum){
        if(idx == len){
            if(sum == target){
                answer++;
            }
            return;
        }
        
        dfs(idx+1, sum + numbers[idx]);
        dfs(idx+1, sum - numbers[idx]);
    }
}